text/3383-recommended-bin-crates.md

```diff
@@ -0,0 +1,162 @@
+- Feature Name: recommended-bin-crates
+- Start Date: 2023-01-04
+- RFC PR: [rust-lang/rfcs#3383](https://github.com/rust-lang/rfcs/pull/3383)
+
+# Summary
+[summary]: #summary
+
+Add an optional `recommended-bin-crates` field to the `[package]`
+section of `Cargo.toml`, to enable crate authors to point out related
+binary crates in the error message Cargo users get when attempting to
+`cargo install` a crate without binaries.
+
+# Motivation
+[motivation]: #motivation
+
+Command-line tools written in Rust are often published in crates named
+different than the command, since that name is already occupied by a
+related library crate, for instance:
+
+* the `diesel` command is provided by the `diesel_cli` binary crate,
+  that depends on the `diesel` library crate
+
+* the `wasm-bindgen` command is provided by the `wasm-bindgen-cli`
+  binary crate, which is different from the `wasm-bindgen` library crate
+
+While such a setup has several benefits, it currently leads to a
+user experience problem with Cargo: To obtain a command, users will be
+tempted to run `cargo install <command>`, which will however inevitably fail:
+
+```
+$ cargo install diesel
+error: there is nothing to install in `diesel v2.0.3`, because it has no binaries
+`cargo install` is only for installing programs, and can't be used with libraries.
+To use a library crate, add it as a dependency in a Cargo project instead.
+```
+
+The idea of this RFC is that the `Cargo.toml` of such
+a library-only crate could specify for instance:
+
+```toml
+[package]
+name = "diesel"
+# ...
+recommended-bin-crates = ["diesel-cli"]
+```
+
+which could be picked up by Cargo in order to additionally include
+a note such as the following in the above error message:
+
+> The developers of `diesel` suggest you may want to install `diesel-cli` instead.
+
+resulting in a more seamless user experience.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following is written as if it was part of the [manifest page] of the Cargo Book.
+
+## The `recommended-bin-crates` field
+
+The `recommended-bin-crates` field is an array of names of related binary crates.
```

Comment:
What if this package, or the other package referred to, is being published to an alternate registry? What are the semantics in that case — does `recommended-bin-crates` always refer to a package in the same registry, or does it always refer to `crates.io`?

Reply:
The intended semantic is that it always refers to packages in the same registry. I'll update the RFC accordingly.